pkg/dns/dns.go

```diff
@@ -0,0 +1,453 @@
+package dns
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/containous/maesh/pkg/k8s"
+	"github.com/containous/traefik/v2/pkg/safe"
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Provider represents a DNS provider.
+type Provider int
+
+// Supported DNS providers.
+const (
+	UnknownDNS Provider = iota
+	CoreDNS
+	KubeDNS
+
+	coreFileHeader  = "#### Begin Maesh Block"
+	coreFileTrailer = "#### End Maesh Block"
+)
+
+var (
+	supportedCoreDNSVersions = []string{
+		"1.3",
+		"1.4",
+		"1.5",
+		"1.6",
+	}
+)
+
+// Client holds the client for interacting with the k8s DNS system.
+type Client struct {
+	client k8s.Client
+	log    logrus.FieldLogger
+}
+
+// NewClient returns an initialized DNSClient object.
+func NewClient(log logrus.FieldLogger, client k8s.Client) *Client {
+	return &Client{
+		client: client,
+		log:    log,
+	}
+}
+
+// CheckDNSProvider checks that the DNS provider that is deployed in the cluster
+// is supported and returns it.
+func (c *Client) CheckDNSProvider() (Provider, error) {
+	c.log.Info("Checking DNS provider")
+
+	match, err := c.coreDNSMatch()
+	if err != nil {
+		return UnknownDNS, err
+	}
+
+	if match {
+		return CoreDNS, nil
+	}
+
+	match, err = c.kubeDNSMatch()
+	if err != nil {
+		return UnknownDNS, err
+	}
+
+	if match {
+		return KubeDNS, nil
+	}
+
+	return UnknownDNS, fmt.Errorf("no supported DNS service available for installing maesh: %w", err)
```

Comment:
For me, it's a bit strange to always return `UnknownDNS` with a corresponding error. 

Is the `UnknownDNS` value really needed? 
Why return an error here if there is no error?

Reply:
Well, the error is that none of the supported DNS services are available, which means Maesh will not work.

If we returned `*Provider` instead, we could return nil instead of `UnknownDNS`. Would that be better?

Reply:
I see, returning `Provider` is fine.